crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,317 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
 const EVP_AEAD *EVP_aead_xchacha20_poly1305(void) {
   return &aead_xchacha20_poly1305;
 }
+
+static int cipher_chacha20_poly1305_init_key(CIPHER_CHACHA_POLY_CTX *ctx,
+                                 const uint8_t user_key[CHACHA_KEY_LEN],
+                                 const uint8_t counter_nonce[CHACHA_CTR_IV_LEN])
+{
+  CIPHER_CHACHA_KEY *key = CC_KEY(ctx);
+  uint32_t i;
+  if (user_key) {
+    for (i = 0; i < (CHACHA_KEY_LEN / 4); i++) {
+      key->key[i] = *(((const uint32_t *) user_key) + i);
+    }
+  }
+  if (counter_nonce) {
+    for (i = 0; i < CHACHA_CTR_IV_LEN / 4; i++) {
+      key->counter_nonce[i] = *(((const uint32_t *) counter_nonce) + i);
+    }
+  }
+  key->partial_len = 0;
+  return 1;
+}
+
+static int cipher_chacha20_poly1305_init(EVP_CIPHER_CTX *ctx,
+                                         const uint8_t *key,
+                                         const uint8_t *iv, int32_t enc) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  if (!key && !iv) {
+    return 1;
```

Comment:
This looks weird but is consistent with other EVP_CIPHER implementations. Is this documented?